File: cpp/toyapp.cpp
#include "camera.hpp"
#include "mocktracker.hpp"
#include "spdlog/spdlog.h"

#include <iterator>
#include <iostream>
#include <sstream>
#include <chrono>

#include <opencv2/imgcodecs.hpp>
#include <opencv2/videoio.hpp>
#include <opencv2/highgui.hpp>

#include "helper.hpp"


namespace spd = spdlog;


void tracking_cb(MockTracker* tracker) {
    auto logger = spd::get("toy");
    cv::Mat* debug_frame = tracker->get_debug_frame();
    cv::Point toy_center = tracker->get_last_toy_center();
    auto code = tracker->get_toy_code();
    logger->info("toy code:{}, code:{}", toy_code_str(code), code.encode());

    cv::imshow("debug frame", *debug_frame);

    direct_pos_t dp = tracker->get_direct_pos();
    logger->info("center:[{},{}], position:[{},{}], direction:{}", toy_center.x, toy_center.y, dp.position.x, dp.position.y, toy_direct_name(dp.direct));

    char key = (char)cv::waitKey(3);
    if (key == 27 || key == 'q' || key == 'Q') {
        tracker->stop_tracking();
    }
}


int main(int argc, char const *argv[]) {
    // auto logger = spd::stdout_color_mt("toy");
    auto logger = spd::rotating_logger_mt("toy", "toy.log", 1048576 * 5, 3);
    logger->set_level(spd::level::info);
    logger->set_pattern("[%Y-%m-%d %H:%M:%S.%e] [%n] [%l] %v");

    // std::string video_src = "0";
    std::string video_src = "../../image/test.avi";
    SimpleCamera camera(video_src);
    camera.start_camera();

    MockTracker tracker(&camera, 30, true);
    tracker.set_tracking_callback(&tracking_cb);
    tracker.track();

    camera.release_camera();
    std::this_thread::sleep_for(std::chrono::seconds(1));
    return 0;
}

File: py/toyapp.py
#!/usr/bin/env python
# coding: utf-8

import cv2

from tracker import ToyTracker
from camera import SimpleCamera


def tracking_callback():
    # frame = tracker.get_frame()
    debug_frame = tracker.get_debug_image()
    # toy_center = tracker.get_last_toy_center()
    toy_colors = tracker.get_toy_colors()

    # cv2.imshow("original frame", frame)
    cv2.imshow("debug frame", debug_frame)
    key = cv2.waitKey(1)
    if key == 27 or key == ord('q'):
        tracker.stop_tracking()

    # print "toy center: {0}".format(toy_center)
    print "toy colors: {0}".format(toy_colors)


if __name__ == "__main__":
    cam = SimpleCamera(video_src=0)
    # cam = SimpleCamera(video_src='../output.avi')
    cam.start_camera()

    tracker = ToyTracker(camera=cam, max_nb_of_centers=10)
    tracker.set_tracking_callback(tracking_callback=tracking_callback)
    tracker.track()

    cam.release_camera()

